teps/00XX-interceptor-plugins.md

```diff
@@ -0,0 +1,423 @@
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+-->
+
+## Design Details
+
+This section describes the two major portions of the proposal -- the changes to
+the Interceptor interface to make it more typed, and the new CRD to
+install/create interceptors.
+
+### HTTP Interface
+
+This diverges from the current HTTP interface that simply forwards the given request and uses HTTP response codes. Instead the proposed way uses typed response bodies for both requests and response (similar to the AdmissionReviewRequest and Response in AdmissionControllers)
+
+
+#### Request
+
+EL Sink sends incoming events to interceptors over HTTP using a typed format:
+
+
+```go
+	type InterceptorRequest struct {
+		// Body is the incoming HTTP event body
+		Body []byte `json:"body,omitempty"`
+		// Header are the headers for the incoming HTTP event
+		Header map[string][]string `json:"header,omitempty"`
+		// EventUrl is the URL of the incoming event
+		EventUrl string `json:"url,omitempty"`
+		// InterceptorParams are the user specified params for the interceptor
+		InterceptorParams map[string]string `json:"interceptor_ params,omitempty"`
+		// EventID is a unique ID assigned by Triggers to each event
+		EventID string `json:"event_id,omitempty"`
+		// TriggerID is of the form namespace/$ns/triggers/$name
+		TriggerID string `json:"trigger_id,omitempty"`
+	}
+
+```
+
+#### Response
+
+Interceptors in turn return a message with a fixed type:
+
+```go
+type InterceptorResponse struct {
+	// Extensions are  additional fields that is added to the interceptor event.
+	// See TEP-0022. Naming TBD.
+  Extensions       map[string]interface{} `json:"extensions,omitempty"`
+  // Continue indicates if the EventListener should continue processing the Trigger or not
+  Continue         bool                   `json:"continue,omitempty"`
+  // Status is an Error status t 
+	Status           grpc.Status            `json:"status,omitempty"`
+}
+```
+
+### Installing Interceptors
+
+Interceptors are installed via the Interceptor CRD. At minimum the spec declares an `address` field that is the URL where the EventListener should send the request to.
+
+The interceptor spec can also contain option parameters fields that declares any parameters that an interceptor requires. Use cases for this include expressions for CEL interceptor, a reference to shared secret for GitHub/GitLab interceptors etc. 
+
+
+```yaml
+apiVersion: triggers.tekton.dev/v1alpha1
+kind: Interceptor
+metadata:
+  name: CEL
+spec:
+  address: "triggers-cel.tekton-pipelines.svc.local/cel" # any resolvable DNS
+  fields:
+    - name: filter
+      required: false
+    - name: overlay
+      required: false
+status:
+   conditions:
+   - type: READY
+     status: true
+     reason: ServiceReady
+```
+
+
+When an Interceptor is created, the Triggers controller will check that the address is valid and reachable and then set a `status.conditions` field of type `Ready`.
```

Comment:
How will we test if the URL is reachable?

Reply:
So, the more I think about the status.conditions.Ready, the more I feel like we should punt on it for v1 and then revisit later.

My current idea was to just send a "health-check" request and expect any non 500 response. It does get a bit complex since each EL will now need an informer for all Interceptors to check when the health check fails. In addition, the controller/reconciler will have to keep making these health check requests at regular intervals. If there is a minor blip, there is still a chance that the EL makes a call to the Interceptor while its down. In short, it feels like a lot to add for not a ton of benefit. What do you think?